src/chrome/chromeConnection.ts

```diff
@@ -128,9 +129,19 @@ export class ChromeConnection implements IObservableEvents<IStepStartedEventsEmi
     /**
      * Attach the websocket to the first available tab in the chrome instance with the given remote debugging port number.
      */
-    public attach(address = '127.0.0.1', port = 9222, targetUrl?: string, timeout?: number, extraCRDPChannelPort?: number): Promise<void> {
-        return this._attach(address, port, targetUrl, timeout, extraCRDPChannelPort)
-            .then(() => { });
+    public attach(address = '127.0.0.1', port = 9222, targetUrl?: string, timeout = ChromeConnection.ATTACH_TIMEOUT, extraCRDPChannelPort?: number, userDataDir?: string): Promise<void> {
+        let selectedTarget: ITarget;
+        return utils.retryAsync(async () => {
+            const launchedPort = (port === 0 && userDataDir) ? await this.getLaunchedPort(address, userDataDir) : port;
```

Comment:
What will happen if someone calls this with port: 0 and no userDataDir?

Reply:
Changed this. We will now throw an error in that case. We _could_ theoretically try to be smart and figure out the userDataDir, but there are so many different possibilities I'd rather not. The only time this would ever be a problem anyway is if the user passes runtimeExecutable with no data dir and port 0.

Reply:
Chrome does specify defaults for the user data dir [here](https://chromium.googlesource.com/chromium/src/+/master/docs/user_data_dir.md#windows).